src/wp-admin/includes/class-rollback-auto-update.php

```diff
@@ -0,0 +1,405 @@
+		remove_action( 'shutdown', array( new WP_Upgrader(), 'delete_temp_backup' ), 100 );
+	}
+
+	/**
+	 * Restart update process for core.
+	 */
+	private function restart_core_updates() {
+		$core_update = find_core_auto_update();
+		if ( $core_update ) {
+			$core_updater = new WP_Automatic_Updater();
+			$core_updater->update( 'core', $core_update );
+		}
+	}
+
+	/**
+	 * Get array of non-fataling auto-updates remaining.
+	 *
+	 * @return array
+	 */
+	private function get_remaining_auto_updates() {
+		if ( empty( $this->handler_args ) ) {
+			return array();
+		}
+
+		// Get array of plugins set for auto-updating.
+		$auto_updates    = (array) get_site_option( 'auto_update_plugins', array() );
+		$current_plugins = array_keys( static::$current->response );
+
+		// Get all auto-updating plugins that have updates available.
+		$current_auto_updates = array_intersect( $auto_updates, $current_plugins );
+
+		// Get array of non-fatal auto-updates remaining.
+		$remaining_auto_updates = array_diff( $current_auto_updates, self::$processed, self::$fatals );
+
+		return $remaining_auto_updates;
+	}
+
+	/**
+	 * Sends an email noting successful and failed updates.
+	 */
+	private function send_update_result_email() {
+		add_filter( 'auto_plugin_theme_update_email', array( $this, 'props' ), 10, 4 );
+		$successful = array();
+		$failed     = array();
+
+		/*
+		 * Using `get_plugin_data()` instead has produced warnings/errors
+		 * as the files may not be in place at this time.
+		 */
+		$plugins = get_plugins();
+
+		foreach ( static::$current->response as $k => $update ) {
+			$item = static::$current->response[ $k ];
+			$name = $plugins[ $update->plugin ]['Name'];
+
+			/*
+			 * This appears to be the only way to get a plugin's older version
+			 * at this stage of an auto-update when not implementing this
+			 * feature directly in Core.
+			 */
+			$current_version = static::$current->checked[ $update->plugin ];
+
+			/*
+			 * The `current_version` property does not exist yet. Add it.
+			 *
+			 * `static::$current->response[ $k ]` is an instance of `stdClass`,
+			 * so this should not fall victim to PHP 8.2's deprecation of
+			 * dynamic properties.
+			 */
+			$item->current_version = $current_version;
+
+			$plugin_result = (object) array(
+				'name' => $name,
+				'item' => $item,
+			);
+
+			$success = array_diff( self::$processed, self::$fatals );
+
+			if ( in_array( $update->plugin, $success, true ) ) {
+				$successful['plugin'][] = $plugin_result;
+				continue;
+			}
+
+			if ( in_array( $update->plugin, self::$fatals, true ) ) {
+				$failed['plugin'][] = $plugin_result;
+			}
+		}
+
+		$automatic_upgrader      = new WP_Automatic_Updater();
+		$send_plugin_theme_email = new \ReflectionMethod( $automatic_upgrader, 'send_plugin_theme_email' );
+		$send_plugin_theme_email->setAccessible( true );
+		$send_plugin_theme_email->invoke( $automatic_upgrader, 'mixed', $successful, $failed );
```

Comment:
```suggestion
		$send_plugin_theme_email = new \ReflectionMethod( $automatic_upgrader, 'send_plugin_theme_email' );
		$send_plugin_theme_email->setAccessible( true );
		$send_plugin_theme_email->invoke( $automatic_upgrader, 'mixed', $successful, $failed );
		$send_plugin_theme_email->setAccessible( false );
```
Reset the accessibility after we're done with this Core class.

Reply:
I don't understand?

Reply:
To avoid using the Reflection API for this bit, I think we might be better duplicating `WP_Automatic_Updater::send_plugin_theme_email()` into this class for merge, and we can explore ways to reduce code duplication later.